File: 20-actors/tazuna/methods/test_teleop_safety.py
"""Tests for the tazuna teleop-safety reasoner (ADR-2606042100). Stdlib + pytest only."""

from __future__ import annotations

import pytest

from teleop_safety import (
    Command,
    ForceClassError,
    Grant,
    NoServerKeyError,
    TransparentForceError,
    admit_session,
    evaluate,
)

AUTHORIZED = Grant(force_class="soft-actuation", force_auth_ref="forceauth:sanae-soft-001",
                   deadman_ms=300, latency_budget_ms=150)


# ── N1: force-class admission ────────────────────────────────────────────────
def test_weaponizable_force_class_is_unrepresentable():
    with pytest.raises(ForceClassError):
        admit_session(Grant(force_class="weaponizable", force_auth_ref="x"))


def test_arbitrary_force_class_refused():
    with pytest.raises(ForceClassError):
        admit_session(Grant(force_class="lethal", force_auth_ref="x"))


@pytest.mark.parametrize("fc", ["observational", "soft-actuation", "powered-actuation"])
def test_admitted_force_classes_pass(fc):
    admit_session(Grant(force_class=fc, force_auth_ref="forceauth:ok"))  # no raise


# ── G3: Transparent Force ────────────────────────────────────────────────────
def test_grant_without_force_auth_ref_refused():
    with pytest.raises(TransparentForceError):
        admit_session(Grant(force_class="soft-actuation", force_auth_ref=""))


def test_actuation_refused_without_force_auth_even_if_signed():
    with pytest.raises(TransparentForceError):
        evaluate(Command("move", member_sig="m:sig"),
                 Grant(force_class="soft-actuation", force_auth_ref=""))


# ── G4: no-server-key ────────────────────────────────────────────────────────
def test_server_signature_always_refused():
    with pytest.raises(NoServerKeyError):
        evaluate(Command("move", member_sig="m:sig", server_sig="s:sig"), AUTHORIZED)


def test_actuation_requires_member_signature():
    with pytest.raises(NoServerKeyError):
        evaluate(Command("move", member_sig=""), AUTHORIZED)


def test_nominal_actuation_is_member_signed_and_passes():
    v = evaluate(Command("move", member_sig="m:sig", observed_latency_ms=40), AUTHORIZED)
    assert v.safe_state == "nominal"
    assert v.actuates is True
    assert v.effective_kind == "move"


# ── G10: soft-RT supervision ─────────────────────────────────────────────────
def test_deadman_lapse_forces_autonomy_fallback_halt():
    v = evaluate(Command("move", member_sig="m:sig", elapsed_since_presence_ms=900), AUTHORIZED)
    assert v.safe_state == "autonomy-fallback"
    assert v.actuates is False
    assert v.effective_kind == "halt"
    assert "deadman" in v.reason


def test_latency_breach_forces_autonomy_fallback_halt():
    v = evaluate(Command("move", member_sig="m:sig", observed_latency_ms=400), AUTHORIZED)
    assert v.safe_state == "autonomy-fallback"
    assert v.actuates is False
    assert v.effective_kind == "halt"
    assert "latency" in v.reason


def test_deadman_takes_priority_over_latency():
    v = evaluate(Command("move", member_sig="m:sig", elapsed_since_presence_ms=900,
                         observed_latency_ms=400), AUTHORIZED)
    assert "deadman" in v.reason


def test_estop_always_honoured_without_signature():
    v = evaluate(Command("estop"), AUTHORIZED)
    assert v.safe_state == "estopped"
    assert v.actuates is False


def test_halt_and_handback_need_no_signature():
    assert evaluate(Command("halt"), AUTHORIZED).effective_kind == "halt"
    assert evaluate(Command("handback"), AUTHORIZED).effective_kind == "handback"


def test_estop_honoured_even_with_breached_supervision():
    # an e-stop must work regardless of deadman/latency state
    v = evaluate(Command("estop", elapsed_since_presence_ms=99999), AUTHORIZED)
    assert v.safe_state == "estopped"


# ── G7: never live at R0 ─────────────────────────────────────────────────────
def test_actuation_is_advisory_only_at_r0():
    # evaluate() returns a verdict; it never performs a live actuation. `actuates` is a permission
    # flag the cell still routes through dry-run + operator/Council gate (G7).
    v = evaluate(Command("manipulate", member_sig="m:sig", observed_latency_ms=10), AUTHORIZED)
    assert v.actuates is True  # permitted...
    # ...but the lexicon pins dryRun const true and outwardGated const true; no live call here.


def test_unknown_command_kind_rejected():
    with pytest.raises(ValueError):
        evaluate(Command("teleport", member_sig="m:sig"), AUTHORIZED)

File: 20-actors/tazuna/methods/test_teleop_safety.clj
#!/usr/bin/env bb
;; Clojure mirror of methods/test_teleop_safety.py — tazuna 手綱 teleop-safety reasoner.
(ns tazuna.methods.test-teleop-safety
  "1:1 port of the teleop_safety.py suite: N1 force-class admission, G3 Transparent
  Force, G4 no-server-key, G10 soft-RT supervision (deadman > latency priority),
  e-stop/halt/handback always honoured, G7 advisory-only, unknown-kind rejection.
  Errors are checked by the ex-info :error tag (≡ the Python exception subclass).

  Run:  bb --classpath 20-actors 20-actors/tazuna/methods/test_teleop_safety.clj"
  (:require [tazuna.methods.teleop-safety :as t]
            [clojure.string :as str]
            [clojure.test :refer [deftest is run-tests]]))

(def AUTHORIZED (t/grant {:force-class "soft-actuation" :force-auth-ref "forceauth:sanae-soft-001"
                          :deadman-ms 300 :latency-budget-ms 150}))

(defn- error-tag
  "Run f; return the :error tag of the ex-info it throws, or nil if it doesn't throw."
  [f]
  (try (f) nil (catch clojure.lang.ExceptionInfo e (:error (ex-data e)))))

;; ── N1: force-class admission ──
(deftest weaponizable-force-class-unrepresentable
  (is (= :force-class (error-tag #(t/admit-session (t/grant {:force-class "weaponizable" :force-auth-ref "x"}))))))

(deftest arbitrary-force-class-refused
  (is (= :force-class (error-tag #(t/admit-session (t/grant {:force-class "lethal" :force-auth-ref "x"}))))))

(deftest admitted-force-classes-pass
  (doseq [fc ["observational" "soft-actuation" "powered-actuation"]]
    (is (nil? (t/admit-session (t/grant {:force-class fc :force-auth-ref "forceauth:ok"}))))))

;; ── G3: Transparent Force ──
(deftest grant-without-force-auth-ref-refused
  (is (= :transparent-force (error-tag #(t/admit-session (t/grant {:force-class "soft-actuation" :force-auth-ref ""}))))))

(deftest actuation-refused-without-force-auth-even-if-signed
  (is (= :transparent-force
         (error-tag #(t/evaluate (t/command "move" {:member-sig "m:sig"})
                                 (t/grant {:force-class "soft-actuation" :force-auth-ref ""}))))))

;; ── G4: no-server-key ──
(deftest server-signature-always-refused
  (is (= :no-server-key
         (error-tag #(t/evaluate (t/command "move" {:member-sig "m:sig" :server-sig "s:sig"}) AUTHORIZED)))))

(deftest actuation-requires-member-signature
  (is (= :no-server-key (error-tag #(t/evaluate (t/command "move" {:member-sig ""}) AUTHORIZED)))))

(deftest nominal-actuation-passes
  (let [v (t/evaluate (t/command "move" {:member-sig "m:sig" :observed-latency-ms 40}) AUTHORIZED)]
    (is (= "nominal" (:safe-state v)))
    (is (true? (:actuates v)))
    (is (= "move" (:effective-kind v)))))

;; ── G10: soft-RT supervision ──
(deftest deadman-lapse-forces-autonomy-fallback-halt
  (let [v (t/evaluate (t/command "move" {:member-sig "m:sig" :elapsed-since-presence-ms 900}) AUTHORIZED)]
    (is (= "autonomy-fallback" (:safe-state v)))
    (is (false? (:actuates v)))
    (is (= "halt" (:effective-kind v)))
    (is (str/includes? (:reason v) "deadman"))))

(deftest latency-breach-forces-autonomy-fallback-halt
  (let [v (t/evaluate (t/command "move" {:member-sig "m:sig" :observed-latency-ms 400}) AUTHORIZED)]
    (is (= "autonomy-fallback" (:safe-state v)))
    (is (= "halt" (:effective-kind v)))
    (is (str/includes? (:reason v) "latency"))))

(deftest deadman-takes-priority-over-latency
  (let [v (t/evaluate (t/command "move" {:member-sig "m:sig" :elapsed-since-presence-ms 900
                                         :observed-latency-ms 400}) AUTHORIZED)]
    (is (str/includes? (:reason v) "deadman"))))

(deftest estop-always-honoured-without-signature
  (let [v (t/evaluate (t/command "estop") AUTHORIZED)]
    (is (= "estopped" (:safe-state v)))
    (is (false? (:actuates v)))))

(deftest halt-and-handback-need-no-signature
  (is (= "halt" (:effective-kind (t/evaluate (t/command "halt") AUTHORIZED))))
  (is (= "handback" (:effective-kind (t/evaluate (t/command "handback") AUTHORIZED)))))

(deftest estop-honoured-even-with-breached-supervision
  (is (= "estopped" (:safe-state (t/evaluate (t/command "estop" {:elapsed-since-presence-ms 99999}) AUTHORIZED)))))

;; ── G7 advisory-only + unknown kind ──
(deftest actuation-is-advisory-only-at-r0
  (is (true? (:actuates (t/evaluate (t/command "manipulate" {:member-sig "m:sig" :observed-latency-ms 10}) AUTHORIZED)))))

(deftest unknown-command-kind-rejected
  (is (= :value-error (error-tag #(t/evaluate (t/command "teleport" {:member-sig "m:sig"}) AUTHORIZED)))))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'tazuna.methods.test-teleop-safety)]
    (System/exit (if (zero? (+ fail error)) 0 1))))
